src/rdkafka_sasl_oauthbearer.c

```diff
@@ -0,0 +1,1006 @@
+ */
+static int rd_kafka_sasl_oauthbearer_client_new (rd_kafka_transport_t *rktrans,
+                                    const char *hostname,
+                                    char *errstr, size_t errstr_size) {
+        struct rd_kafka_sasl_oauthbearer_state *state;
+
+        state = rd_calloc(1, sizeof(*state));
+        state->state = RD_KAFKA_SASL_OAUTHB_STATE_SEND_CLIENT_FIRST_MESSAGE;
+
+        /*
+         * Save off the state structure now, before any possibility of
+         * returning, so that we will always free up the allocated memory in
+         * rd_kafka_sasl_oauthbearer_close().
+         */
+        rktrans->rktrans_sasl.state = state;
+
+        /*
+         * Make sure we have a consistent view of the token and extensions
+         * throughout the authentication process -- even if it is refreshed
+         * midway through this particular authentication.
+         */
+        rd_kafka_rdlock(rktrans->rktrans_rkb->rkb_rk);
+        if (!rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->token_value) {
+                rd_snprintf(errstr, errstr_size,
+                        "OAUTHBEARER cannot log in because there is no token "
+                        "available; last error: %s",
+                        rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->errstr ?
+                        rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->errstr :
+                        "not available");
+                rd_kafka_rdunlock(rktrans->rktrans_rkb->rkb_rk);
+                return -1;
+        }
+        state->token_value = strdup(
+                rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->token_value);
+        state->md_principal_name = strdup(
+                rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->md_principal_name);
+        rd_list_copy_to(&state->extensions,
+                &rktrans->rktrans_rkb->rkb_rk->rk_oauthbearer->extensions,
+                rd_strtup_list_copy, NULL);
+        rd_kafka_rdunlock(rktrans->rktrans_rkb->rkb_rk);
+
+        /* Kick off the FSM */
+        return rd_kafka_sasl_oauthbearer_fsm(rktrans, NULL, errstr, errstr_size);
+}
+
+
+
+/**
+ * @brief Validate OAUTHBEARER config, which is a no-op
+ * (we rely on initial token retrieval)
+ */
+static int rd_kafka_sasl_oauthbearer_conf_validate (rd_kafka_t *rk,
+                                              char *errstr,
+                                              size_t errstr_size) {
+        /*
+         * We will rely on the initial token retrieval as a proxy
+         * for configuration validation.
+         */
+        return 0;
+}
+
+
+
+
+const struct rd_kafka_sasl_provider rd_kafka_sasl_oauthbearer_provider = {
+        .name          = "OAUTHBEARER (builtin)",
+        .client_new    = rd_kafka_sasl_oauthbearer_client_new,
+        .recv          = rd_kafka_sasl_oauthbearer_recv,
+        .close         = rd_kafka_sasl_oauthbearer_close,
+        .conf_validate = rd_kafka_sasl_oauthbearer_conf_validate,
+};
```

Comment:
After this, add a unit test for all the JWS parsing, since string handling is very error prone in C.

Do this:
 * add a `int unittest_sasl_oauthbroker (void) ..`  (see example here https://github.com/edenhill/librdkafka/blob/master/src/rdbuf.c#L1541)
 * add declaration to rdkafka_sasloau..h
 * in rdunittest.c: include rdkafka_sasl..h in rdunittest.c and add "sasloauthbearer" to the unittests array in rd_unittest.c
 * run the unit tests by `make 0000` in tests/

Reply:
👍 Done